Async project goal

nikomatsakis opened the pull request:
This is a proposed flagship goal for 2024h2 covering **async Rust**. You can [read more about the project goal slate and its associated process here](https://rust-lang.github.io/rust-project-goals/2024h2/slate.html). This RFC is prepared using the project goal template, which differs from the typical RFC template.


The overall goal is **bringing the Async Rust experience closer to parity with sync Rust**. We have identified three high-priority goals that we believe would do the most to improve async over the long term:

* resolving the "send bound problem", thus enabling foundational, generic traits like Tower's [`Service`]() trait;
* stabilizing async closures, thus enabling richer, combinator APIs like sync Rust's [`Iterator`](https://doc.rust-lang.org/std/iter/trait.Iterator.html);
* reorganizing the async WG, so the project can benefit from a group of async rust experts with deep knowledge of the space that can align around a shared vision;

We have also identified two "stretch goals":

* stabilizing a trait in libstd for async iteration, thus enabling the ecosystem to build atop a stable foundation;
* completing the async drop experiments proposed in [MCP 727][], laying the groundwork for resolving the the next major gap in language feature support.

Approving this goal implies agreement from the [Lang][], [Libs][], and [Libs-API][] teams to the items marked as ![Team][] in the table of work items, along with potentially other design meetings as needed.

[Team]: https://img.shields.io/badge/Team%20ask-red
[MCP 727]: https://github.com/rust-lang/compiler-team/issues/727
[Lang]: https://www.rust-lang.org/governance/teams/lang
[Libs]: https://www.rust-lang.org/governance/teams/library
[Libs-API]: https://www.rust-lang.org/governance/teams/library#team-libs-api


[Rendered](https://github.com/nikomatsakis/rfcs/blob/async-project-goal/text/0000-async-project-goal-2024h2.md)

lovely-error commented:
I think all goals you described can be achieved by implementing these three features:

1. Since some people say that generisizing futures would lead to monomoprhisation bloat rust need [unsized values](https://github.com/nikomatsakis/rfcs/blob/async-project-goal/text/1909-unsized-rvalues.md)
2. [Couroutines](https://github.com/nikomatsakis/rfcs/blob/async-project-goal/text/2033-experimental-coroutines.md) subsume everything async does and they also subsume effect systems. I know from experience that they also do async better than current rust async.
3. Async drop won't be needed if rust gets `no leak` types.

nikomatsakis commented:
@rfcbot fcp merge

This project goal has been under development and discussion for some time. Since discussion on the thread has been relatively quiet, I'm going to go ahead and propose to merge it now. There will be a long list of checkboxes regardless so I want to get that discussion up and going.

tmandry commented on text/0000-async-project-goal-2024h2.md:
```diff
@@ -0,0 +1,275 @@
+
+### First focus: language parity, interop traits
+
+Based on the above analysis, the Rust org has been focused on driving async/sync language parity, especially in those areas that block the development of a rich ecosystem. The biggest progress took place in [Dec 2023][afitblog], when async fn in traits and return position impl trait in trait were stabilizd. Other work includes authoring the async vision doc, stabilizing helpers like [`std::future ...
```
For vision doc I assume you mean the original one, which is easy to confuse with the vision RFC being proposed here.

nikomatsakis replied:
I updated your comment to clarify better

tmandry commented on text/0000-async-project-goal-2024h2.md:
```diff
@@ -0,0 +1,275 @@
+
+### Stabilize async closures
+
+Building ergonomic APIs in async is often blocked by the lack of *async closures*. Async combinator-like APIs today typically make use an ordinary Rust closure that returns a future, such as the `filter` API from [`StreamExt`](https://docs.rs/futures/latest/futures/prelude/stream/trait.StreamExt.html#method.filter):
```
```suggestion
Building ergonomic APIs in async is often blocked by the lack of *async closures*. Async combinator-like APIs today typically make use of an ordinary Rust closure that returns a future, such as the `filter` API from [`StreamExt`](https://docs.rs/futures/latest/futures/prelude/stream/trait.StreamExt.html#method.filter):
```

tmandry commented on text/0000-async-project-goal-2024h2.md:
```diff
@@ -0,0 +1,275 @@
+
+### Resolve the ["send bound"][sb] problem
+
+Although async functions in traits were stabilized, there is currently no way to write a generic function that requires only impls where the returned futures are `Send`. This blocks the use of async function in traits in some core ecosystem crates, such as [tower](https://crates.io/crates/tower), which want to work across all kinds of async executors. This problem is called the ["send bound"][sb] ...
```
```suggestion
Although async functions in traits were stabilized, there is currently no way to write a generic function that requires impls where the returned futures are `Send`. This blocks the use of async function in traits in some core ecosystem crates, such as [tower](https://crates.io/crates/tower), which want to work across all kinds of async executors. This problem is called the ["send bound"][sb] problem and there has been extensive discussion of the various ways to solve it. [RFC #3654][] has been opened proposing one solution and describing why that path is preferred. Our goal for the year is to adopt *some* solution on stable.
```

tmandry commented on text/0000-async-project-goal-2024h2.md:
```diff
@@ -0,0 +1,275 @@
+
+[AVD]: https://rust-lang.github.io/wg-async/vision.html
+
+This RFC will lay out a "plan of attack" for async, including both obvious good things (similar to [async closures][]) but also "known unknowns" and ways to resolve them. Areas the RFC is expected to cover are as follows:
```
```suggestion
This RFC will lay out a "plan of attack" for async, including both obvious good things (similar to [async closures](#stabilize-async-closures)) but also "known unknowns" and ways to resolve them. Areas the RFC is expected to cover are as follows:
```

tmandry commented on text/0000-async-project-goal-2024h2.md:
```diff
@@ -0,0 +1,275 @@
+
+### Complete async drop experiments
+
+Authors of async code frequently need to call async functions as part of resource cleanup. Because Rust today only supports synchronous destructors, this cleanup must take place using alternative mechanisms, forcing a divergence between sync Rust (which uses destructors to arrange cleanup) and async Rust. [MCP 727][] proposed a series of experiments aimed at supporting async drop in the compiler. ...
```
```suggestion
Authors of async code frequently need to call async functions as part of resource cleanup. Because Rust today only supports synchronous destructors, this cleanup must take place using alternative mechanisms, forcing a divergence between sync Rust (which uses destructors to arrange cleanup) and async Rust. [MCP 727][] proposed a series of experiments aimed at supporting async drop in the compiler. We would like to continue and complete those experiments. These experiments are aimed at defining how support for async drop will be implemented in the compiler and some possible ways that we could modify the type system to support it (in particular, one key question is how to prevent types whose drop is async from being dropped in sync code).
```

tmandry commented on text/0000-async-project-goal-2024h2.md:
```diff
@@ -0,0 +1,275 @@
+    * Common idioms contain footguns that cause unexpected failures (see e.g., Tyler's blog post [Making Async Rust Reliable][])
+    * Low-level performance hurdles, such as large future sizes and downsides of the poll model
+    * Fragmentation between runtimes
+* Design axioms to pursue for async (see e.g. axioms proposed)
```
```suggestion
* Design axioms to pursue for async (see e.g. axioms proposed below)
```

tmandry commented on text/0000-async-project-goal-2024h2.md:
```diff
@@ -0,0 +1,275 @@
+    * Low-level performance hurdles, such as large future sizes and downsides of the poll model
+    * Fragmentation between runtimes
+* Design axioms to pursue for async (see e.g. axioms proposed)
+* Goals, some variant of
```
How do you think these goals should relate to the "shiny future" and axioms proposed in this goal RFC? Are the goals based on those, but slightly more specific, or with concrete and measurable outcomes?

nikomatsakis replied:
Good question. I imagined these goals as being roughly equivalent to the shiny-future shown here, I think. It's possible we should remove some of this outline, but I did want to use it to advertise and drive alignment early ont the expected "rough direction".

nikomatsakis replied:
I decided to just "redirect" to the shiny future

tmandry commented on text/0000-async-project-goal-2024h2.md:
```diff
@@ -0,0 +1,275 @@
+
+# Design axiom
+
+* **We lay the foundations for a thriving ecosystem.** The role of the Rust org is to deelop the rudiments that support an interoperable and thriving async crates.io ecosystem.
```
```suggestion
* **We lay the foundations for a thriving ecosystem.** The role of the Rust org is to develop the rudiments that support an interoperable and thriving async crates.io ecosystem.
```

tmandry commented on text/0000-async-project-goal-2024h2.md:
```diff
@@ -0,0 +1,275 @@
+# Design axiom
+
+* **We lay the foundations for a thriving ecosystem.** The role of the Rust org is to deelop the rudiments that support an interoperable and thriving async crates.io ecosystem.
+* **Uphold sync's Rust bar for reliability.** Sync Rust famously delivers on the general feeling of "if it compiles, in works" -- async Rust should do the same.
```
```suggestion
* **Uphold sync Rust's bar for reliability.** Sync Rust famously delivers on the general feeling of "if it compiles, in works" -- async Rust should do the same.
```

tmandry commented on text/0000-async-project-goal-2024h2.md:
```diff
@@ -0,0 +1,275 @@
+
+Expectation is that 3-4 design meetings will be needed from lang over the course of H2 and 1-2 from libs API. Reviewing the async vision doc is expected to be the biggest requirement.
+
+# Outputs and milestones
```
This section seems incomplete.. but also a bit redundant.

nikomatsakis replied:
Yes, it should just be removed.

tmandry commented on text/0000-async-project-goal-2024h2.md:
```diff
@@ -0,0 +1,275 @@
+
+There has been extensive discussion about the best design for the "Stream" or "async iter" trait and we judge that the design space is well understood. We would like to unblock generator syntax in 2025 which will require some form of trait.
+
+The majority of the debate about the trait has been on the topic of whether to base the trait on a `poll_next` function, as we do today, or to try and make the trait use `async fn next`, making it more anaologous with the `Iterator` trait (and potentially even making it be two versions of a single trait). We will definitely explore forwards compatibility questions as part of this discussion. niko ...
```
```suggestion
The majority of the debate about the trait has been on the topic of whether to base the trait on a `poll_next` function, as we do today, or to try and make the trait use `async fn next`, making it more anaologous with the `Iterator` trait (and potentially even making it be two versions of a single trait). We will definitely explore forwards compatibility questions as part of this discussion. nikomatsakis for example still wants to explore maybe-async-like designs, especially for combinator APIs like `map`. However, we also refer to the design axiom that "when in doubt, zero-cost is our compass" -- we believe we should be able to stabilize a trait that does the low-level details right, and then design higher level APIs atop that.
```

tmandry commented on text/0000-async-project-goal-2024h2.md:
```diff
@@ -0,0 +1,275 @@
+
+## What about "maybe async", effect systems, and keyword generics?
+
+Keyword generics is an ambitious initiative to enable code that is "maybe async". It has generated significant controversy, with some people feeling it is necessary for Rust to scale and others judging it to be overly complex. We anticipate having more debate on this topic as part of drafting the async vision doc.
```
I don't want to set an expectation that we will settle this question fully.. but I think we can put some parameters around it.

nikomatsakis replied:
Update.

nikomatsakis commented on text/0000-async-project-goal-2024h2.md:
```diff
@@ -0,0 +1,275 @@
+
+* **Rust scales up**. Async Rust reduces cost for large dataplanes because a single server can serve high load without significantly increasing tail latency.
+* **Rust scales down.** Async Rust can be run without requiring a garbage collector or [even an operating system][embassy], making it a great fit for embedded systems.
+* **Rust is reliable.** Networked services run 24/7, so Rust's "if it compiles, it works" mantra means unexpected failures and, in turn, fewer pages in the middle of the night.
```
lol

nikomatsakis reviewed (commented):
Thanks for the edits. Will respond now to the more thoughtful questions.

pachi commented on text/0000-async-project-goal-2024h2.md:
```diff
@@ -0,0 +1,279 @@
+# Design axiom
+
+* **We lay the foundations for a thriving ecosystem.** The role of the Rust org is to develop the rudiments that support an interoperable and thriving async crates.io ecosystem.
+* **Uphold sync Rust's bar for reliability.** Sync Rust famously delivers on the general feeling of "if it compiles, in works" -- async Rust should do the same.
```
Typo ->If It compiles, it works

nikomatsakis replied:
Thanks! Fixed.

nikomatsakis commented:
After discussion with @tmandry I have made two major updates to the goal structure

* First, we removed the async vision doc, which seemed a bit premature and overly ambitious. We have replaced it with the goal to reorganize the async WG. 
* Second, I added more explicit prioritization, clarifying the async iteration + async drop are "stretch goals". The prioritization is based on what will set us up for long-term success.

nikomatsakis commented:
Upon reflection, I've decided to change the process for approving project goals. I'm going to close this RFC, but the content is still correct.

bew commented:
Hello, I was following this rfc (was probably mentioned in TWiR), is there a new/different issue/object to follow? 🤔

clarfonthey commented:
I shared a similar question on the other RFC that was closed relating to project goals; might be worth centralising the discussion there: https://github.com/rust-lang/rfcs/pull/3658#issuecomment-2192460574

Although presumably an alternative link to follow this specific goal will be provided here once that discussion has finished.
